### ISSUE TITLE
zpaq algorithm does not appear to match impl in zpaq upstream

### ISSUE THREAD

**Issue by jmesmon** [remram44/cdchunking-rs#6]
Hi, I was looking at the chunking provided by Zpaq in this crate (as part of comparing it to a zpaq implimentation I've written), and noticed that it doesn't appear to match up with the upstream zpaq code. 

In particular, it appears that the hash update equation doesn't match up (though I might be missing some transform you've done).

[cdchunking code](https://github.com/remram44/cdchunking-rs/blob/6870ff693f1f505abb92c83c0d98401f849ebce5/src/lib.rs#L496-L500):
```
const HM: Wrapping<u32> = Wrapping(123_456_791);
...
pub fn update(&mut self, byte: u8) -> bool {
        if byte == self.o1[self.c1 as usize] {
            self.h = self.h * HM + Wrapping(byte as u32 + 1);
        } else {
            self.h = self.h * HM * Wrapping(2) + Wrapping(byte as u32 + 1);
        }
        self.o1[self.c1 as usize] = byte;
        self.c1 = byte;

        self.h.0 < (1 << self.nbits)
}
```

[zpaq code](https://github.com/zpaq/zpaq/blob/9ab539f644e364f0d92e2918b90ce2534c75653f/zpaq.cpp#L2403-L2410)
```
            if (c==o1[c1]) h=(h+c+1)*314159265u, ++hits;
            else h=(h+c+1)*271828182u;
            o1[c1]=c;
            c1=c;
```

In particular, zpaq adds `h`, `c`, and `1` prior to multiplying (and uses different constants), but cdchunking adds `c` and `1` after `h` is mulitiplied.

I've done some slight modification to zpaq itself so that it prints chunk lengths, and I'm not seeing a correspondence between the lengths produced by zpaq and cdchunking.

```
diff --git a/zpaq.cpp b/zpaq.cpp
index f468a7b..84e1f5f 100644
--- a/zpaq.cpp
+++ b/zpaq.cpp
@@ -83,6 +83,7 @@ Possible options:
 #endif
 #ifdef unix
 #define PTHREAD 1
+#include <inttypes.h>
 #include <sys/param.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -2416,6 +2417,8 @@ int Jidac::add() {
         assert(sz<=MAX_FRAGMENT);
         total_done+=sz;
 
+       printf("## %" PRIu64 "\n", sz);
+
         // Look for matching fragment
         assert(uint64_t(sz)==sha1.usize());
         memcpy(sha1result, sha1.result(), 20);
```

- Is this difference in algorithm an oversight?
- Or is it purposefully different?
 - If it is purposefully different, is it following another already published source?
 - If it is purposefully different, could the divergence be documented?

**Comment by remram44** [comment]
Honestly I cannot remember where I got the code. I don't think it was the official libzpaq source.

The difference is unintended, and I think we should use the official algorithm since it is named in the API. That would be a breaking change, but I don't know if there are any real users of this crate (I wrote it mostly for a proof of concept, I never got around to the project that was supposed to use it).

I suppose the crate should also have test vectors generated from the C++ implementation.

**Comment by remram44** [comment]
Specification appears under "Implementation notes" in [zpaq206.pdf](http://mattmahoney.net/dc/zpaq206.pdf)

@jmesmon Do you have test vectors that you have checked with the C++ implementation or your own?

**Comment by jmesmon** [comment]
I've only generated 2 test vectors, both use 32KiB of data generated with Pcg64 as the input and a list of chunk lengths as the output.

```
generator: Pcg64::new(0, 0xa02bdbf7bb3c0a7ac28fa16a64abf96)
zpaq-fragment: 3
length: 32768
cuts: [10785, 6329, 1287, 860, 4716, 7419]
```

```
generator: Pcg64::new(3, 0xa02bdbf7bb3c0a7ac28fa16a64abf96)
zpaq-fragment: 3
length: 32768
cuts: [16353, 2334, 970, 5326, 1557]
```

note:
  - that the really long number passed to Pcg64 is the specified default Pcg64 stream, I'm just tweaking the seed
  - zpaq (the c++ code) "fragment: 3" corresponds to "ZPAQ::new(13)" in your implementation (zpaq c++ works in KiB)
  - I've used the same length for both of these

The code exists in [this test in this branch of my repo](https://github.com/jmesmon/hash-roll/blob/956f80acf1868caf9eae6896ab98815b54b43255/tests/cuts.rs#L154-L180).

That branch also has code to generate a binary file that can be passed to zpaq.

**Comment by remram44** [comment]
I made the change and imported your tests as best I could, but [the results don't match exactly](https://github.com/remram44/cdchunking-rs/runs/991761075?check_suite_focus=true) (left is `cdchunking`, right is `hash-roll`):

```
---- tests::test_long_0 stdout ----
thread 'tests::test_long_0' panicked at 'assertion failed: `(left == right)`
  left: `[10785, 6329, 1287, 860, 97, 4619, 7419, 1372]`,
 right: `[10785, 6329, 1287, 860, 4716, 7419]`', src/lib.rs:659:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::test_long_3 stdout ----
thread 'tests::test_long_3' panicked at 'assertion failed: `(left == right)`
  left: `[16353, 2334, 970, 5326, 1557, 6228]`,
 right: `[16353, 2334, 970, 5326, 1557]`', src/lib.rs:676:9
```

The last number is probably because you're only listing the cut points and not all chunk sizes, but this `97` is suspicious.

**Comment by jmesmon** [comment]
yep, minimum chunk size is likely the distinction here. zpaq c++ includes a minimum and maximum chunk size. zpaq c++ calculates it from the selected average chunk size (ie: "bits", 13, etc).

https://github.com/zpaq/zpaq/blob/9ab539f644e364f0d92e2918b90ce2534c75653f/zpaq.cpp#L2411-L2414

```
          if (c==EOF
              || sz>=MAX_FRAGMENT
              || (fragment<=22 && h<(1u<<(22-fragment)) && sz>=MIN_FRAGMENT))
            break;
```

https://github.com/zpaq/zpaq/blob/9ab539f644e364f0d92e2918b90ce2534c75653f/zpaq.cpp#L2198-L2201
```
  const unsigned MAX_FRAGMENT=fragment>19 || (8128u<<fragment)>blocksize-12
      ? blocksize-12 : 8128u<<fragment;
  const unsigned MIN_FRAGMENT=fragment>25 || (64u<<fragment)>MAX_FRAGMENT
      ? MAX_FRAGMENT : 64u<<fragment;
```

(imo: the calculations here are pretty unreadable)

**Comment by remram44** [comment]
I added a warning and released 1.0.1. I don't think it's worth matching the official constants, and this crate is unused as far as I know.

You seem to be doing a much more conscientious job with [`hash-roll`](https://github.com/jmesmon/hash-roll), which should probably be used instead of cdchunking if people want those well-known algorithm you implemented to specs (although your `Splitter` trait doesn't seem written for streaming or memory-constrained operations, it needs to collect a whole chunk in a `Vec`).

**Comment by jmesmon** [comment]
Understood, thats fine with me.

On hash-roll: the stuff in `master` is pretty old. the `next` branch is a complete rewrite which adjusts the apis a bit (and entirely deletes `Splitter`). It does include some apis that support some streaming type operations, though they need a bit more polish (imo).

The [some of the APIs](https://docs.rs/cdchunking/1.0.1/cdchunking/#from-a-file-object-streaming-chunks-with-zero-allocation) you've written for cdchunking are things I'm keeping in mind as I'm building easier-to-use abstractions over the current low-level hash-roll::next APIs (like `fn find_chunk_edge(&self, state: &mut Self::SearchState, data: &[u8]) -> (Option<usize>, usize);`. No one really _wants_ to use an api like that)
The API you've used for
